### DISCUSSION TITLE
Date mocking does not work in V9

### DISCUSSION THREAD

**Discussion by meelis-krasmus** [gpbl/react-day-picker#2472] [Support] [answered]
Hello
I am unable to mock dates with [MockDate](https://www.npmjs.com/package/mockdate) after migrating to V9. 
This is necessary for testing purposes. I understand that this functionality can be replaced using the `defaultMonth` prop, but I wanted to inform you just in case.

**Comment by gpbl** [comment]
@melkras some help reproducing your issue with MockDate? Why are you unable to use it?

**Reply by meelis-krasmus** [reply]
When rendering `<DayPicker />` in the testing environment and setting a mocked date with `MockDate.set('2024-07-01')`, the mocked date wont affect `DayPicker's` currently visible month as it used to in V8.
When mocking it should render the currently visible month of July 2024, instead it renders September 2024.
So I'm assuming that for some reason the `DayPicker's` internal Date.now() cannot be overriden like this anymore.

**Reply by leonaves** [reply]
Hey, I also ran into this problem after updating and I think I've found the cause. Because of how the new custom dateLib functionality works, what's happening is that RDP defaults to using the system Date function here:

https://github.com/gpbl/react-day-picker/blob/b5e554af2be050e68276bfd97e7e01667e3123ea/src/lib/dateLib.ts#L54

And then that object reference is captured within this closure:

https://github.com/gpbl/react-day-picker/blob/9c45c9f698eea439cc722a10413f95bf41d43dfd/src/helpers/getDateLib.ts

Meaning that later when trying to work out "today", because RDP uses `new datelib.Date()` here:

https://github.com/gpbl/react-day-picker/blob/b5e554af2be050e68276bfd97e7e01667e3123ea/src/helpers/getInitialMonth.ts#L24

It uses that original Date reference. This will break any mocking library that replaces the global Date object after everything has been imported, which is pretty much how every library that does date mocking works. A workaround is to supply a `dateLib` property of:

```jsx
<ReactDayPicker dateLib={{ Date }} />
```

When rendering.

**Reply by gpbl** [reply]
Good find @leonaves! A straight solution there could be to update the line above:

```ts
'dateLib' in props ? new dateLib.Date() : new Date(), 
```

Anyway, I could mock dates in Jest: https://github.com/gpbl/react-day-picker/blob/main/examples/Start.test.tsx#L14-L17

**Comment by gpbl** [comment, marked as answer]
Mocking date should be fixed in v9.1.2 - please write back if you are still encountering issues mocking the dates, thanks!
